client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,343 @@
+    hand.forEach((order) => {
+      const card = deck[order];
+      if (
+        shouldCalculateCard(card, playerIndex, ourPlayerIndex, deck, oldDeck)
+      ) {
+        cardsToCalculate.push(card);
+      }
+    });
+  });
+  return cardsToCalculate;
+}
+
+function shouldCalculateCard(
+  card: CardState,
+  playerIndex: number,
+  ourPlayerIndex: number,
+  deck: CardState[],
+  oldDeck: readonly CardState[],
+) {
+  const cardPossibilitiesForPlayer = getCardPossibilitiesForPlayer(
+    card,
+    playerIndex,
+    ourPlayerIndex,
+  );
+
+  if (
+    card.revealedToPlayer[playerIndex] ||
+    cardPossibilitiesForPlayer.length === 1
+  ) {
+    // The player already knows what this card is.
+    return false;
+  }
+
+  const oldCard = oldDeck[card.order];
+
+  if (typeof oldCard === "undefined" || oldCard.location === "deck") {
+    // this is a newly drawn card and hasn't had any calculations yet.
+    return true;
+  }
+
+  // If the possibilities on the unknown cards don't change, then the result of our calculation
+  // won't change.  We only need to recalculate the card if the input (possibilities) changed.
+  return unknownCardPossibilitiesDifferent(
+    card.order,
+    deck,
+    oldDeck,
+    playerIndex,
+    ourPlayerIndex,
+  );
+}
+
+function getCardPossibilitiesForPlayer(
+  card: CardState,
+  playerIndex: number,
+  ourPlayerIndex: number,
+): ReadonlyArray<readonly [number, number]> {
+  if (card.location === playerIndex) {
+    // If this card is in the players hand, then use our best (empathy) guess.
+    return card.possibleCardsForEmpathy;
+  }
+  if (playerIndex === ourPlayerIndex || card.revealedToPlayer[playerIndex]) {
+    // This is revealed to the player or we are the requested player => just use our best knowledge.
+    return card.possibleCards;
+  }
+  // This is an unrevealed card not in the players hand but not revealed to them.
+  // That can happen with something like a detrimental character (such as 'Slow-Witted')
+  // or throw it in the hole.  We can't use our best (empathy) guess, because it might be in our own
+  // hand and we might know more about the card then the other player does.  We know the other
+  // player at least knows about the clues for it, so we'll use that set of possibilities.
+  return card.possibleCardsFromClues;
+}
+
+function generatePossibilitiesForUnknownCards(
+  excludeCardOrder: number,
+  deck: readonly CardState[],
+  playerIndex: number,
+  ourPlayerIndex: number,
+): ReadonlyArray<ReadonlyArray<readonly [number, number]>> {
+  const unknownCards = generateUnknownCards(
+    excludeCardOrder,
+    deck,
+    playerIndex,
+  );
+  const possibilities: Array<ReadonlyArray<readonly [number, number]>> = [];
+  for (const card of unknownCards) {
+    possibilities.push(
+      getCardPossibilitiesForPlayer(card, playerIndex, ourPlayerIndex),
+    );
+  }
+  // start with the more stable possibilities
+  possibilities.sort((a, b) => a.length - b.length);
```

Comment:
Why does this matter?

Reply:
This is for performance. It seemed to have a measurable difference.

The possibilityValid method will short-circuit if it finds a branch that's impossible or if it finds a possibility that's valid. Here's an attempt at an example:

possibilities = [[red 5 or yellow 5],[green or blue],[green or blue],[green or blue],[red 5]]

possibilityValid would initially start with the first card being red 5. It would then check about 1000 combinations of the next three cards before finding each one is impossible at the very end of each combination. If we reorder that to:

possibilities = [[red 5],[red 5 or yellow 5],[green or blue],[green or blue],[green or blue]]

Then when it attempts to resolve [red 5 or yellow 5] to red 5, it will realize that's impossible and short-circuit that branch (not checking the next 1000 combinations of the next 3 cards). It would switch to checking if the combination would work when [red 5 or yellow 5] resolves to yellow 5 (by finding a combination of the next 3 cards that fit).

So it should get to a fitting combination quicker or find that there is no fitting combination quicker.

Reply:
This applies to more than just cards that have one possibility (such as red 5 in the previous example).

Reply:
i mean all of what you wrote should be sticked into a comment above the code in question then

Reply:
Yeah, that's fair. This assumes that I've explained it well enough here.

Reply:
Oh this makes sense especially now that I understand the comparison function. I thought `a.length - b.length` was trying to compare how the number of possibilities had changed since last deduction but it's just comparing two elements of the same list

Reply:
I added this as a comment to the code